docs/features/source-generators.md

```diff
@@ -0,0 +1,285 @@
+
+```csharp
+    public struct SourceGeneratorContext
+    {
+        public Compilation Compilation { get; }
+
+        // TODO: replace AnalyzerOptions with an differently named type that is otherwise identical.
+        // The concern being that something added to one isn't necessarily applicable to the other.
+        public AnalyzerOptions AnalyzerOptions { get; }
+
+        public CancellationToken CancellationToken { get; }
+
+        public void ReportDiagnostic(Diagnostic diagnostic) { throw new NotImplementedException(); }
+
+        public void AddSource(string fileNameHint, SourceText sourceText) { throw new NotImplementedException(); }
+    }
+}
+```
+It is assumed that some generators will want to generate more than one `SourceText`, for example in a 1:1 mapping 
+for additional files. The `fileNameHint` parameter of `AddSource` is intended to address this:
+
+1. If the generated files are emitted to disk, having some ability to put some distinguishing text might be useful. 
+For example, if you have two `.resx` files, generating the files with simply names of `ResxGeneratedFile1.cs` and
+`ResxGeneratedFile2.cs` wouldn't be terribly useful -- you'd want it to be something like
+`ResxGeneratedFile-Strings.cs` and `ResxGeneratedFile-Icons.cs` if you had two `.resx` files
+ named "Strings" and "Icons" respectively.
+
+2. The IDE needs some concept of a "stable" identifier. Source generators create a couple of fun problems for the IDE: 
+users will want to be able to set breakpoints in a generated file, for example. If a source generator outputs multiple 
+files we need to know which is which so we can know which file the breakpoints go with. A source generator of course is 
+allowed to stop emitting a file if needs change (if you delete a `.resx`, then the generated file associated with it 
+will also go away), but this gives us some control here.
+
+This was called "hint" in that the compiler is implicitly allowed to control the filename in however it ultimately 
+needs, and if two source generators give the same "hint" it can still distinguish them with any sort of 
+prefix/suffix as necessary.
+
+### IDE Integration
+
+One of the more complicated aspects of supporting generators is enabling a high-fidelity 
+experience in Visual Studio. For the purposes of determining code correctness, it is 
+expected that all generators will have had to be run. Obviously, it is impractical to run 
+every generator on every keystroke, and still maintain an acceptable level of performance 
+within the IDE.
+
+#### Progressive complexity opt-in
+
+It is expected instead that source generators would work on an 'opt-in' approach to IDE 
+enablement. 
+
+By default, a generator implementing only `ISourceGenerator` would see no IDE integration
+and only be correct at build time. Based on conversations with 1st party customers, 
+there are several cases where this would be enough.
```

Comment:
This does not instill confidence. I would prefer to have a default of "runs as a background operation after a source file is saved", similar to an asynchronous version of single file generators. If we want to improve performance, there are other good approaches we can use, such as:

1. Filtering (e.g. by input file name or by an attribute applied within a file).
2. Design-time generation of a reference API, separate from implementation generation. This could save time through reduced validation of inputs and avoiding generation of member implementations.

Reply:
There are lots of use cases for generators where there is no user-observable output from the source generator (think statically compiled DI etc). We want to start with the absolute simplest model, and build up the complexity from there.

---
In reply to: [354526436](https://github.com/dotnet/roslyn/pull/40162#discussion_r354526436) [](ancestors = 354526436)